### PR TITLE
BUG: Enforce high >= low on uniform number generators

### LINKED ISSUE #17905: np.random.uniform does not raise ValueError when `low > high`, even though other distribution do (like np.random.randint)
`print(np.random.randint(2, 1))` raises `ValueError`, also the documentation of `np.random.uniform` says those inputs are `low` and `high`.

### Reproducing code example:

```python
import numpy as np
print(np.random.uniform(2, 1))
```

### NumPy/Python version information:

1.19.2 3.7.7 (default, May  7 2020, 21:25:33) 
[GCC 7.3.0]

### PR DESCRIPTION
**Pull request by bashtage**
Check that high is weakly larger than low and raise if now

closes #17905

### CODE REVIEW

**Comment by rkern** [comment]
I don't think this qualifies for an exception to the freezing of `RandomState` per NEP 19. I think this should be confined to `Generator`.

Please update the `Notes` section which still refers to the old behavior.

**Comment by bashtage** [comment]
I agree with @rkern that we cannot change mtrand.pyx in this instance, and so the PR is limited to the new generator.  It adds a check.  An alternative would be to better explain that role of high low and make this the default behaviour.  Something along the lines of

```
The input values are not checked for order. The uniform is generated as low + (high - low) * random(), so that if
high < low, then the uniform is generated as if high and low are swaped.
```

Probably simpler to enforce though.  Probably will also help to avoid hard to detect bugs by end-users.

File: numpy/random/tests/test_generator_mt19937.py [hunk 1]
```diff
@@ -1666,6 +1666,12 @@ def test_uniform_range_bounds(self):
         # DBL_MAX by increasing fmin a bit
         random.uniform(low=np.nextafter(fmin, 1), high=fmax / 1e17)
 
+    def test_uniform_neg_range(self):
+        func = random.uniform
+        assert_raises(ValueError, func, 2, 1)
+        assert_raises(ValueError, func,  [1, 2], [1, 1])
+        assert_raises(ValueError, func,  [[0, 1],[2, 3]], 2)
```
Comment (seberg) [inline numpy/random/tests/test_generator_mt19937.py:1673]:
```suggestion
        assert_raises(ValueError, func,  [[0, 1], [2, 3]], 2)
```

**Review by seberg — COMMENTED** [review]
Thanks, I am happy to put this in then. Two small things:
1. I guess it may be worth adding a release note, since it is a compatibility change in theory? (even if the new API isn't very old anyway)
2. Just curious what happens/should happen for `low == high`, and if we should test that.

**Comment by rkern** [comment]
> 2\. Just curious what happens/should happen for `low == high`, and if we should test that.

That case is documented, if not tested. It always returns the constant `low` (or equivalently, `high`).

**Comment by bashtage** [comment]
It is also mathematically well defined using a Dirac delta.

**Comment by charris** [comment]
What about `integers`? There is #14333, which has other problems, but it seems somewhat related.

**Comment by bashtage** [comment]
> What about `integers`? There is #14333, which has other problems, but it seems somewhat related.

#14333 is mostly a preference about the error message. This is about whether this is high > low is actually an error.

**Comment by mattip** [comment]
Thanks @bashtage

**Comment by eric-wieser** [comment]
With this change, it's now harder to generate a random number in the half-open interval `(low, high]`, which I get the feeling was what passing the arguments in the "wrong" order did. Are we sure it wasn't better to support that use-case?

**Comment by bashtage** [comment]
> With this change, it's now harder to generate a random number in the half-open interval `(low, high]`, which I get the feeling was what passing the arguments in the "wrong" order did. Are we sure it wasn't better to support that use-case?

That feels like a magic incarnation, at least given the lack of clear intent for this case doc. It doesn't seem to be that much larger to do what you wish.

```
-gen.uniform(-high, -low)
```

**Comment by rkern** [comment]
With floating point arithmetic, there are very no real guarantees that you'll actually get a half-open interval, on either side.

**Comment by eric-wieser** [comment]
> there are very no real guarantees that you'll actually get a half-open interval, on either side.

Ah, I was misled by the docstring which seems to suggest the interval is truly half-open, only to remark in the notes that this is not guaranteed. Sounds like my concern isn't relevant, thanks for clarifying.

**Comment by rkern** [comment]
There's definitely room for a `uniform`-like method that fully controls whether the ends are open or closed, in any combination. But it should also enforce `low <= high` and control the open/closed-ness through other arguments.

**Pull request merged**